servicetalk-capacity-limiter-api/build.gradle

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: "io.servicetalk.servicetalk-gradle-plugin-internal-library"
+
+dependencies {
+  implementation platform(project(":servicetalk-dependencies"))
+  api project(":servicetalk-concurrent-api")
+  api project(":servicetalk-transport-api")
```

Comment:
This modules depends only on `:servicetalk-context-api`.

Here is my patch for some other minor comments:
```patch
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpServiceFilter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpServiceFilter.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpServiceFilter.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpServiceFilter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpServiceFilter.java	(date 1715210265384)
@@ -127,12 +127,12 @@
     }
 
     @Override
-    protected Ticket wrapTicket(@Nullable final ServerListenContext serverListenContext, final Ticket ticket) {
+    Ticket wrapTicket(@Nullable final ServerListenContext serverListenContext, final Ticket ticket) {
         return serverListenContext == null ? ticket : new ServerResumptionTicketWrapper(serverListenContext, ticket);
     }
 
     @Override
-    protected Single<StreamingHttpResponse> handleLocalCapacityRejection(
+    Single<StreamingHttpResponse> handleLocalCapacityRejection(
             @Nullable final ServerListenContext serverListenContext,
             final StreamingHttpRequest request,
             @Nullable final StreamingHttpResponseFactory responseFactory) {
@@ -154,7 +154,7 @@
     }
 
     @Override
-    protected Single<StreamingHttpResponse> handleLocalBreakerRejection(
+    Single<StreamingHttpResponse> handleLocalBreakerRejection(
             final StreamingHttpRequest request,
             @Nullable final StreamingHttpResponseFactory responseFactory,
             @Nullable final CircuitBreaker breaker) {
@@ -392,7 +392,6 @@
          * @return {@code this}.
          */
         public Builder observer(final TrafficResiliencyObserver observer) {
-            requireNonNull(observer, "observer");
             this.observer = new SafeTrafficResiliencyObserver(observer);
             return this;
         }
@@ -644,6 +643,7 @@
             this.listenContext = listenContext;
         }
 
+        @Nullable
         @Override
         public CapacityLimiter.LimiterState state() {
             return ticket.state();
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java	(date 1715210265378)
@@ -170,7 +170,6 @@
             // reactive flow isn't followed) we still complete ticket lifetime.
             try {
                 final TicketObserver ticketObserver = observer.onAllowedThrough(request, ticket.state());
-                assert ticketObserver != null;
                 return handleAllow(delegate, request, wrapTicket(serverListenContext, ticket), ticketObserver,
                         breaker, startTime).shareContextOnSubscribe();
             } catch (Throwable cause) {
@@ -180,16 +179,16 @@
         });
     }
 
-    protected Ticket wrapTicket(@Nullable final ServerListenContext serverListenContext, final Ticket ticket) {
+    Ticket wrapTicket(@Nullable final ServerListenContext serverListenContext, final Ticket ticket) {
         return ticket;
     }
 
-    protected abstract Single<StreamingHttpResponse> handleLocalCapacityRejection(
+    abstract Single<StreamingHttpResponse> handleLocalCapacityRejection(
             @Nullable ServerListenContext serverListenContext,
             StreamingHttpRequest request,
             @Nullable StreamingHttpResponseFactory responseFactory);
 
-    protected abstract Single<StreamingHttpResponse> handleLocalBreakerRejection(
+    abstract Single<StreamingHttpResponse> handleLocalBreakerRejection(
             StreamingHttpRequest request,
             @Nullable StreamingHttpResponseFactory responseFactory,
             @Nullable CircuitBreaker breaker);
@@ -245,18 +244,14 @@
                             }
                         } finally {
                             onSuccessTicketTerminal.accept(ticket);
-                            if (ticketObserver != null) {
-                                ticketObserver.onComplete();
-                            }
+                            ticketObserver.onComplete();
                         }
                     }
 
                     @Override
                     public void onError(final Throwable throwable) {
                         AbstractTrafficManagementHttpFilter.this.onError(throwable, breaker, startTimeNs, ticket);
-                        if (ticketObserver != null) {
-                            ticketObserver.onError(throwable);
-                        }
+                        ticketObserver.onError(throwable);
                     }
 
                     @Override
@@ -267,9 +262,7 @@
                             }
                         } finally {
                             onCancellationTicketTerminal.accept(ticket);
-                            if (ticketObserver != null) {
-                                ticketObserver.onCancel();
-                            }
+                            ticketObserver.onCancel();
                         }
                     }
                 }, true));
@@ -318,6 +311,7 @@
             this.requestHashCode = requestHashCode;
         }
 
+        @Nullable
         @Override
         public CapacityLimiter.LimiterState state() {
             return delegate.state();
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/SafeTrafficResiliencyObserver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/SafeTrafficResiliencyObserver.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/SafeTrafficResiliencyObserver.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/SafeTrafficResiliencyObserver.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/SafeTrafficResiliencyObserver.java	(date 1715207765167)
@@ -26,6 +26,7 @@
 import javax.annotation.Nullable;
 
 import static io.servicetalk.apple.traffic.resilience.http.NoOpTrafficResiliencyObserver.NO_OP_TICKET_OBSERVER;
+import static java.util.Objects.requireNonNull;
 
 final class SafeTrafficResiliencyObserver implements TrafficResiliencyObserver {
 
@@ -33,7 +34,7 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(SafeTrafficResiliencyObserver.class);
 
     SafeTrafficResiliencyObserver(final TrafficResiliencyObserver original) {
-        this.original = original;
+        this.original = requireNonNull(original);
     }
 
     @Override
@@ -65,7 +66,6 @@
         }
     }
 
-    @Nullable
     @Override
     public TicketObserver onAllowedThrough(final StreamingHttpRequest request,
                                            @Nullable final CapacityLimiter.LimiterState state) {
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrackPendingRequestsHttpFilter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrackPendingRequestsHttpFilter.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrackPendingRequestsHttpFilter.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrackPendingRequestsHttpFilter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrackPendingRequestsHttpFilter.java	(date 1715207314786)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2023 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/PeerCapacityRejectionPolicy.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/PeerCapacityRejectionPolicy.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/PeerCapacityRejectionPolicy.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/PeerCapacityRejectionPolicy.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/PeerCapacityRejectionPolicy.java	(date 1715207298573)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2022 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilter.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilter.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilter.java	(date 1715210265380)
@@ -116,7 +116,7 @@
     /**
      * Default rejection observer for dropped requests from an external sourced.
      * see. {@link Builder#peerCapacityRejection(PeerCapacityRejectionPolicy)}.
-     *
+     * <p>
      * The default predicate matches the following HTTP response codes:
      * <ul>
      *     <li>{@link HttpResponseStatus#TOO_MANY_REQUESTS}</li>
@@ -135,7 +135,7 @@
     /**
      * Default rejection observer for dropped requests from an external sourced due to service unavailability.
      * see. {@link Builder#peerBreakerRejection(HttpResponseMetaData, CircuitBreaker)}}.
-     *
+     * <p>
      * The default predicate matches the following HTTP response codes:
      * <ul>
      *     <li>{@link HttpResponseStatus#SERVICE_UNAVAILABLE}</li>
@@ -197,14 +197,14 @@
     }
 
     @Override
-    protected Single<StreamingHttpResponse> handleLocalBreakerRejection(
+    Single<StreamingHttpResponse> handleLocalBreakerRejection(
             StreamingHttpRequest request, @Nullable StreamingHttpResponseFactory responseFactory,
             @Nullable CircuitBreaker breaker) {
         return DEFAULT_BREAKER_REJECTION;
     }
 
     @Override
-    protected Single<StreamingHttpResponse> handleLocalCapacityRejection(
+    Single<StreamingHttpResponse> handleLocalCapacityRejection(
             @Nullable final ServerListenContext serverListenContext,
             StreamingHttpRequest request, @Nullable StreamingHttpResponseFactory responseFactory) {
         return RETRYABLE_LOCAL_CAPACITY_REJECTION;
@@ -276,7 +276,7 @@
          * @param capacityLimiterSupplier The {@link Supplier} to create a new {@link CapacityLimiter} for each new
          * filter created by this {@link StreamingHttpClientFilterFactory factory}.
          */
-        public Builder(Supplier<CapacityLimiter> capacityLimiterSupplier) {
+        public Builder(final Supplier<CapacityLimiter> capacityLimiterSupplier) {
             requireNonNull(capacityLimiterSupplier);
             this.capacityPartitionsSupplier = () -> {
                 CapacityLimiter capacityLimiter = capacityLimiterSupplier.get();
@@ -450,8 +450,7 @@
          * peer-rejection or not.
          * @return {@code this}.
          */
-        public Builder peerUnavailableRejectionPredicate(
-                final Predicate<HttpResponseMetaData> rejectionPredicate) {
+        public Builder peerUnavailableRejectionPredicate(final Predicate<HttpResponseMetaData> rejectionPredicate) {
             this.peerUnavailableRejectionPredicate = requireNonNull(rejectionPredicate);
             return this;
         }
@@ -531,7 +530,6 @@
          * @return {@code this}.
          */
         public Builder observer(final TrafficResiliencyObserver observer) {
-            requireNonNull(observer);
             this.observer = new SafeTrafficResiliencyObserver(observer);
             return this;
         }
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiters.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiters.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiters.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiters.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiters.java	(date 1715203237425)
@@ -44,12 +44,12 @@
      * Returns a {@link CapacityLimiter} that will reject all requests till the current pending request count is equal
      * or less to the passed {@code capacity}.
      * This {@link CapacityLimiter} takes into consideration the {@link Classification} of a given request and will
-     * variate the effective {@code capacity} according to the {@link Classification#priority() weight} before
+     * variate the effective {@code capacity} according to the {@link Classification#priority() priority} before
      * attempting to grant access to the request. The effective {@code capacity} will never be more than the given
      * {@code capacity}.
      * <p>
-     * Requests with {@link Classification#priority() weight} equal to or greater than {@code 100} will enjoy
-     * the full capacity (100%), while requests with {@link Classification#priority() weight} less than {@code 100}
+     * Requests with {@link Classification#priority() priority} equal to or greater than {@code 100} will enjoy
+     * the full capacity (100%), while requests with {@link Classification#priority() priority} less than {@code 100}
      * will be mapped to a percentage point of the given {@code capacity} and be granted access only if the {@code
      * consumed capacity} is less than that percentage.
      * <br>
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterUtils.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterUtils.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterUtils.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterUtils.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterUtils.java	(date 1715203146995)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2022 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-traffic-resilience-http/gradle/spotbugs/test-exclusions.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/gradle/spotbugs/test-exclusions.xml b/servicetalk-traffic-resilience-http/gradle/spotbugs/test-exclusions.xml
--- a/servicetalk-traffic-resilience-http/gradle/spotbugs/test-exclusions.xml	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/gradle/spotbugs/test-exclusions.xml	(date 1715207281774)
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-  ~ Copyright © 2019 Apple Inc. and the ServiceTalk project authors
+  ~ Copyright © 2024 Apple Inc. and the ServiceTalk project authors
   ~
   ~ Licensed under the Apache License, Version 2.0 (the "License");
   ~ you may not use this file except in compliance with the License.
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiterBuilder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiterBuilder.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiterBuilder.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiterBuilder.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiterBuilder.java	(date 1715206352242)
@@ -15,16 +15,14 @@
  */
 package io.servicetalk.apple.capacity.limiter.api;
 
-import io.servicetalk.utils.internal.DurationUtils;
-import io.servicetalk.utils.internal.NumberUtils;
-
 import java.time.Duration;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.LongSupplier;
-import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
+import static io.servicetalk.utils.internal.DurationUtils.ensureNonNegative;
 import static io.servicetalk.utils.internal.NumberUtils.ensureBetweenZeroAndOneExclusive;
+import static io.servicetalk.utils.internal.NumberUtils.ensurePositive;
 import static io.servicetalk.utils.internal.NumberUtils.ensureRange;
 import static java.lang.Integer.MAX_VALUE;
 import static java.util.Objects.requireNonNull;
@@ -138,7 +136,7 @@
      * @return {@code this}.
      */
     public AimdCapacityLimiterBuilder increment(final float increment) {
-        this.increment = NumberUtils.ensurePositive(increment, "increment");
+        this.increment = ensurePositive(increment, "increment");
         return this;
     }
 
@@ -151,7 +149,7 @@
      * @return {@code this}.
      */
     public AimdCapacityLimiterBuilder cooldown(final Duration duration) {
-        this.cooldown = DurationUtils.ensureNonNegative(duration, "cooldown");
+        this.cooldown = ensureNonNegative(duration, "cooldown");
         return this;
     }
 
@@ -190,9 +188,8 @@
                 timeSource);
     }
 
-    @Nonnull
     private String name() {
-        return name == null ? AimdCapacityLimiter.class.getSimpleName() + "_" + SEQ_GEN.incrementAndGet() : name;
+        return name == null ? AimdCapacityLimiter.class.getSimpleName() + '-' + SEQ_GEN.incrementAndGet() : name;
     }
 
     /**
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AllowAllCapacityLimiter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AllowAllCapacityLimiter.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AllowAllCapacityLimiter.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AllowAllCapacityLimiter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AllowAllCapacityLimiter.java	(date 1715203146993)
@@ -50,6 +50,7 @@
     };
 
     private AllowAllCapacityLimiter() {
+        // Singleton
     }
 
     @Override
@@ -64,6 +65,6 @@
 
     @Override
     public String toString() {
-        return AllowAllCapacityLimiter.class.getSimpleName();
+        return name();
     }
 }
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiter.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiter.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/AimdCapacityLimiter.java	(date 1715203146993)
@@ -52,7 +52,7 @@
  * represent a constant rate (i.e., has no notion of time).
  * <p>
  * The solution is based on the <a href="https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease">
- * AIMD feedback control algorithm</a>
+ * AIMD feedback control algorithm</a>.
  */
 final class AimdCapacityLimiter implements CapacityLimiter {
 
@@ -214,7 +214,6 @@
             this.remaining = remaining;
         }
 
-        @Nullable
         @Override
         public LimiterState state() {
             return this;
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiter.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiter.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiter.java	(date 1715206352245)
@@ -61,9 +61,9 @@
     }
 
     @Override
-    public Ticket tryAcquire(final Classification classification, final ContextMap meta) {
-        final int weight = min(max(classification.priority(), 0), 100);
-        final int effectiveLimit = (capacity * weight) / 100;
+    public Ticket tryAcquire(final Classification classification, @Nullable final ContextMap meta) {
+        final int priority = min(max(classification.priority(), 0), 100);
+        final int effectiveLimit = (capacity * priority) / 100;
         for (;;) {
             final int currPending = pending;
             if (currPending == effectiveLimit &&
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CompositeCapacityLimiter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CompositeCapacityLimiter.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CompositeCapacityLimiter.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CompositeCapacityLimiter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CompositeCapacityLimiter.java	(date 1715206352249)
@@ -20,6 +20,8 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import javax.annotation.Nullable;
+
 import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.joining;
 
@@ -32,23 +34,24 @@
 final class CompositeCapacityLimiter implements CapacityLimiter {
 
     private final List<CapacityLimiter> providers;
-    private final String namesCsv;
+    private final String name;
 
     CompositeCapacityLimiter(final List<CapacityLimiter> providers) {
         if (requireNonNull(providers).isEmpty()) {
             throw new IllegalArgumentException("Empty capacity limiters.");
         }
         this.providers = new ArrayList<>(providers);
-        this.namesCsv = providers.stream().map(CapacityLimiter::name).collect(joining(", "));
+        this.name = CompositeCapacityLimiter.class.getSimpleName() + "[ " +
+                providers.stream().map(CapacityLimiter::name).collect(joining(", ")) + " ]";
     }
 
     @Override
     public String name() {
-        return CompositeCapacityLimiter.class.getSimpleName() + "[ " + namesCsv + " ]";
+        return name;
     }
 
     @Override
-    public Ticket tryAcquire(final Classification classification, final ContextMap context) {
+    public Ticket tryAcquire(final Classification classification, @Nullable final ContextMap context) {
         Ticket[] results = null;
         int idx = 0;
         for (CapacityLimiter provider : providers) {
@@ -72,7 +75,7 @@
         return compositeResult(results);
     }
 
-    private int completed(Ticket[] results) {
+    private static int completed(Ticket[] results) {
         int remaining = 1;
         for (Ticket ticket : results) {
             if (ticket == null) {
@@ -86,7 +89,7 @@
         return remaining;
     }
 
-    private int failed(Throwable cause, Ticket[] results) {
+    private static int failed(Throwable cause, Ticket[] results) {
         int remaining = 1;
         for (Ticket ticket : results) {
             if (ticket == null) {
@@ -100,7 +103,7 @@
         return remaining;
     }
 
-    private int dropped(Ticket[] results) {
+    private static int dropped(Ticket[] results) {
         int remaining = 1;
         for (Ticket ticket : results) {
             if (ticket == null) {
@@ -114,7 +117,7 @@
         return remaining;
     }
 
-    private int cancelled(Ticket[] results) {
+    private static int cancelled(Ticket[] results) {
         int remaining = 1;
         for (Ticket ticket : results) {
             if (ticket == null) {
@@ -128,33 +131,33 @@
         return remaining;
     }
 
-    private Ticket compositeResult(final Ticket[] tickets) {
+    private static Ticket compositeResult(final Ticket[] tickets) {
         return new Ticket() {
             @Override
             public LimiterState state() {
-                // Targeting the most specific one (assuming an order of rate-limiter, customer-quota-limiter
+                // Targeting the most specific one (assuming an order of rate-limiter, customer-quota-limiter).
                 // In the future we could make this configurable if proven useful.
                 return tickets[tickets.length - 1].state();
             }
 
             @Override
             public int completed() {
-                return CompositeCapacityLimiter.this.completed(tickets);
+                return CompositeCapacityLimiter.completed(tickets);
             }
 
             @Override
             public int failed(final Throwable cause) {
-                return CompositeCapacityLimiter.this.failed(cause, tickets);
+                return CompositeCapacityLimiter.failed(cause, tickets);
             }
 
             @Override
             public int dropped() {
-                return CompositeCapacityLimiter.this.dropped(tickets);
+                return CompositeCapacityLimiter.dropped(tickets);
             }
 
             @Override
             public int ignored() {
-                return CompositeCapacityLimiter.this.cancelled(tickets);
+                return CompositeCapacityLimiter.cancelled(tickets);
             }
         };
     }
Index: servicetalk-circuit-breaker-api/build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-circuit-breaker-api/build.gradle b/servicetalk-circuit-breaker-api/build.gradle
--- a/servicetalk-circuit-breaker-api/build.gradle	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-circuit-breaker-api/build.gradle	(date 1715206550272)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-circuit-breaker-api/src/main/java/io/servicetalk/apple/circuit/breaker/api/package-info.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-circuit-breaker-api/src/main/java/io/servicetalk/apple/circuit/breaker/api/package-info.java b/servicetalk-circuit-breaker-api/src/main/java/io/servicetalk/apple/circuit/breaker/api/package-info.java
--- a/servicetalk-circuit-breaker-api/src/main/java/io/servicetalk/apple/circuit/breaker/api/package-info.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-circuit-breaker-api/src/main/java/io/servicetalk/apple/circuit/breaker/api/package-info.java	(date 1715206550275)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/NoOpTrafficResiliencyObserver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/NoOpTrafficResiliencyObserver.java b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/NoOpTrafficResiliencyObserver.java
--- a/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/NoOpTrafficResiliencyObserver.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/NoOpTrafficResiliencyObserver.java	(date 1715207765153)
@@ -45,7 +45,6 @@
                                       final ContextMap meta, final Classification classification) {
     }
 
-    @Nullable
     @Override
     public TicketObserver onAllowedThrough(final StreamingHttpRequest request,
                                            @Nullable final CapacityLimiter.LimiterState state) {
Index: servicetalk-capacity-limiter-api/build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/build.gradle b/servicetalk-capacity-limiter-api/build.gradle
--- a/servicetalk-capacity-limiter-api/build.gradle	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/build.gradle	(date 1715207247786)
@@ -18,18 +18,15 @@
 
 dependencies {
   implementation platform(project(":servicetalk-dependencies"))
-  api project(":servicetalk-concurrent-api")
-  api project(":servicetalk-transport-api")
+  api project(":servicetalk-context-api")
 
   implementation project(":servicetalk-annotations")
-  implementation project(":servicetalk-concurrent-internal")
   implementation project(":servicetalk-utils-internal")
   implementation "com.google.code.findbugs:jsr305"
   implementation "org.slf4j:slf4j-api"
 
   testImplementation enforcedPlatform("org.junit:junit-bom:$junit5Version")
   testImplementation testFixtures(project(":servicetalk-concurrent-internal"))
-  testImplementation project(":servicetalk-concurrent-test-internal")
   testImplementation project(":servicetalk-test-resources")
   testImplementation "org.junit.jupiter:junit-jupiter-api"
   testImplementation "org.hamcrest:hamcrest"
Index: servicetalk-traffic-resilience-http/src/test/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilterTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-traffic-resilience-http/src/test/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilterTest.java b/servicetalk-traffic-resilience-http/src/test/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilterTest.java
--- a/servicetalk-traffic-resilience-http/src/test/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilterTest.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-traffic-resilience-http/src/test/java/io/servicetalk/apple/traffic/resilience/http/TrafficResilienceHttpClientFilterTest.java	(date 1715207290908)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2022 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-circuit-breaker-resilience4j/gradle/spotbugs/test-exclusions.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-circuit-breaker-resilience4j/gradle/spotbugs/test-exclusions.xml b/servicetalk-circuit-breaker-resilience4j/gradle/spotbugs/test-exclusions.xml
--- a/servicetalk-circuit-breaker-resilience4j/gradle/spotbugs/test-exclusions.xml	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-circuit-breaker-resilience4j/gradle/spotbugs/test-exclusions.xml	(date 1715206864072)
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-  ~ Copyright © 2019 Apple Inc. and the ServiceTalk project authors
+  ~ Copyright © 2024 Apple Inc. and the ServiceTalk project authors
   ~
   ~ Licensed under the Apache License, Version 2.0 (the "License");
   ~ you may not use this file except in compliance with the License.
Index: servicetalk-circuit-breaker-resilience4j/build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-circuit-breaker-resilience4j/build.gradle b/servicetalk-circuit-breaker-resilience4j/build.gradle
--- a/servicetalk-circuit-breaker-resilience4j/build.gradle	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-circuit-breaker-resilience4j/build.gradle	(date 1715207247788)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -19,8 +19,8 @@
 dependencies {
   implementation platform(project(":servicetalk-dependencies"))
 
-  api "io.github.resilience4j:resilience4j-circuitbreaker:${resilience4jVersion}"
   api project(":servicetalk-circuit-breaker-api")
+  api "io.github.resilience4j:resilience4j-circuitbreaker:${resilience4jVersion}"
 
   implementation project(":servicetalk-annotations")
 
Index: servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/package-info.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/package-info.java b/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/package-info.java
--- a/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/package-info.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/package-info.java	(date 1715206864075)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiterBuilder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiterBuilder.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiterBuilder.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiterBuilder.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/FixedCapacityLimiterBuilder.java	(date 1715203324661)
@@ -16,7 +16,6 @@
 package io.servicetalk.apple.capacity.limiter.api;
 
 import java.util.concurrent.atomic.AtomicInteger;
-import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 import static java.util.Objects.requireNonNull;
@@ -46,7 +45,7 @@
     /**
      * Defines the fixed capacity for the {@link CapacityLimiter}.
      * Concurrent requests above this figure will be rejected. Requests with particular
-     * {@link Classification#priority() weight} will be respected and the total capacity for them will be adjusted
+     * {@link Classification#priority() priority} will be respected and the total capacity for them will be adjusted
      * accordingly.
      * @param capacity The max allowed concurrent requests that this {@link CapacityLimiter} should allow.
      * @return {@code this}.
@@ -81,9 +80,8 @@
         return new FixedCapacityLimiter(name(), capacity, observer);
     }
 
-    @Nonnull
     private String name() {
-        return name == null ? FixedCapacityLimiter.class.getSimpleName() + "_" + SEQ_GEN.incrementAndGet() : name;
+        return name == null ? FixedCapacityLimiter.class.getSimpleName() + '-' + SEQ_GEN.incrementAndGet() : name;
     }
 
     /**
Index: servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/Resilience4jAdapters.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/Resilience4jAdapters.java b/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/Resilience4jAdapters.java
--- a/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/Resilience4jAdapters.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-circuit-breaker-resilience4j/src/main/java/io/servicetalk/apple/capacity/management/resilience4j/Resilience4jAdapters.java	(date 1715206864069)
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -65,8 +65,8 @@
         private R4jCircuitBreaker(final io.github.resilience4j.circuitbreaker.CircuitBreaker breaker) {
             this.breaker = breaker;
             this.breaker.getEventPublisher().onStateTransition(event -> {
-                if (STATE_TRANSITION.equals(event.getEventType()) &&
-                        OPEN.equals(event.getStateTransition().getToState())) {
+                if (event.getEventType() == STATE_TRANSITION &&
+                        event.getStateTransition().getToState() == OPEN) {
                         breakerOpenedAtMillis = NANOSECONDS.toMillis(nanoTime());
                 }
             });
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/Classification.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/Classification.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/Classification.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/Classification.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/Classification.java	(date 1715203192346)
@@ -17,6 +17,7 @@
 
 /**
  * Classification of requests.
+ * <p>
  * In the context of capacity, classification can be used to allow prioritization of requests under
  * certain conditions. When a system is load-shedding, it can still choose to accommodate important demand.
  * The classification is not a feature supported by all {@link CapacityLimiter}s but rather the
@@ -34,8 +35,10 @@
     /**
      * The priority should be a positive number between 0 and 100 (inclusive), which hints to a {@link CapacityLimiter}
      * the importance of a {@link Classification}.
+     * <p>
      * Higher value represents the most important {@link Classification}, while lower value represents less important
      * {@link Classification}.
+     *
      * @return A positive value between 0 and 100 (inclusive) that hints importance of a request to a
      * {@link CapacityLimiter}.
      */
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiter.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiter.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/CapacityLimiter.java	(date 1715203146994)
@@ -52,7 +52,6 @@
  * {@link Classification} is used as hint from the user of the importance of the incoming request, and are not
  * guaranteed to have an influence to the decision if the {@link CapacityLimiter} doesn't support them or chooses to
  * ignore them.
- *
  */
 public interface CapacityLimiter {
 
@@ -67,7 +66,7 @@
      * {@link ContextMap context}.
      *
      * @param classification A class tha represents the <i>importance</i> of a request, to be evaluated upon permit.
-     * @param context Contextual metadata supported for evaluation from the call-site. This, in an Http context
+     * @param context Contextual metadata supported for evaluation from the call-site. This, in an HTTP context
      * would typically be the {@code HttpRequest#context()}.
      * @return {@link Ticket} when capacity is enough to satisfy the demand or {@code null} when not.
      * @see CapacityLimiter
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiter.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiter.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiter.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiter.java	(date 1715206352256)
@@ -41,6 +41,8 @@
 import java.util.function.BiPredicate;
 import java.util.function.LongSupplier;
 
+import javax.annotation.Nullable;
+
 import static java.lang.Double.isNaN;
 import static java.lang.Math.max;
 import static java.lang.Math.min;
@@ -122,7 +124,7 @@
     }
 
     @Override
-    public Ticket tryAcquire(final Classification classification, final ContextMap meta) {
+    public Ticket tryAcquire(final Classification classification, @Nullable final ContextMap meta) {
         int newPending;
         int newLimit;
 
@@ -153,7 +155,7 @@
         // When positive gradient, and limit already above initial,
         // avoid increasing the limit when we are far from meeting it - i.e. blast radius.
         final boolean isPositiveSuspended = !isNaN(gradient) &&
-                (gradient > 1.0 && limit > initial && suspendLimitInc.test(pending, limit));
+                gradient > 1.0 && limit > initial && suspendLimitInc.test(pending, limit);
         // When negative gradient, and consumption not close to limit,
         // avoid decreasing the limit. Low RPS & noisy environments (RTT deviations > 500ms) tend to bring the limit
         // down to "min" even though there aren't many pending requests to justify the decision.
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/LatencyTracker.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/LatencyTracker.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/LatencyTracker.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/LatencyTracker.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/LatencyTracker.java	(date 1715206352247)
@@ -31,6 +31,7 @@
  * tracker, allowing a way to enhance the algorithm's behavior according to the observations.
  * Implementations must provide thread-safety guarantees.
  */
+@FunctionalInterface
 interface LatencyTracker {
 
     /**
Index: servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterBuilder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterBuilder.java b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterBuilder.java
--- a/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterBuilder.java	(revision baac8d66a661ae6878454fb7ea05afda7ecbc35a)
+++ b/servicetalk-capacity-limiter-api/src/main/java/io/servicetalk/apple/capacity/limiter/api/GradientCapacityLimiterBuilder.java	(date 1715206352252)
@@ -27,7 +27,6 @@
 import java.util.function.BiPredicate;
 import java.util.function.LongSupplier;
 import java.util.function.Predicate;
-import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 import static io.servicetalk.apple.capacity.limiter.api.GradientCapacityLimiterProfiles.DEFAULT_INITIAL_LIMIT;
@@ -176,8 +175,7 @@
      * @return {@code this}.
      */
     public GradientCapacityLimiterBuilder headroom(final BiFunction<Double, Double, Double> headroom) {
-        requireNonNull(headroom);
-        this.headroom = headroom;
+        this.headroom = requireNonNull(headroom);
         return this;
     }
 
@@ -243,8 +241,7 @@
      * @return {@code this}.
      */
     public GradientCapacityLimiterBuilder minGradient(final float minGradient) {
-        ensureBetweenZeroAndOne(minGradient, "minGradient");
-        this.minGradient = minGradient;
+        this.minGradient = ensureBetweenZeroAndOne(minGradient, "minGradient");;
         return this;
     }
 
@@ -258,8 +255,7 @@
      * @return {@code this}.
      */
     public GradientCapacityLimiterBuilder maxGradient(final float maxPositiveGradient) {
-        ensureGreaterThan(maxPositiveGradient, 1.0f, "maxGradient");
-        this.maxGradient = maxPositiveGradient;
+        this.maxGradient = ensureGreaterThan(maxPositiveGradient, 1.0f, "maxGradient");
         return this;
     }
 
@@ -327,9 +323,8 @@
                 suspendLimitInc, suspendLimitDec, headroom, timeSource);
     }
 
-    @Nonnull
     private String name() {
-        return name == null ? GradientCapacityLimiter.class.getSimpleName() + "_" + SEQ_GEN.incrementAndGet() : name;
+        return name == null ? GradientCapacityLimiter.class.getSimpleName() + '-' + SEQ_GEN.incrementAndGet() : name;
     }
 
     /**

```

Reply:
Opened a PR to apply this patch: #2915